protocols/direct-funding.go

```diff
@@ -0,0 +1,198 @@
+package protocols
+
+import (
+	"math/big"
+
+	"github.com/statechannels/go-nitro/channel/state"
+	"github.com/statechannels/go-nitro/types"
+)
+
+// TODO this protocol does not specify how events are handled at all
+// (it assumes that events are handled by pushing information into the store)
+
+// Crank inspects the extended state and declares a list of Effects to be executed
+// It's like a state machine transition function where the finite / enumerable state is returned (computed from the extended state)
+// rather than being independent of the extended state; and where there is only one type of event ("the crank") with no data on it at all
+func (s DirectFundingObjectiveState) Crank(secretKey *[]byte) (Objective, SideEffects, WaitingFor, error) {
+
+	// Input validation
+	if s.Status != Approved {
+		return s, NoSideEffects, WaitingForNothing, ErrNotApproved
```

Comment:
We should return a clone of `s`, I think.

(Even if cloning is trivial, it's nice to have it written out explicitly).

The pattern I anticipate is to clone `s` immediately, and then go about mutating the clone before returning it.